### PR TITLE
750 dynamic permission followup

### LINKED ISSUE #750: Dynamic permission follow up
Follow up for https://github.com/openmsupply/open-msupply/pull/611

- [ ] simplify permissions -> reduce number of program related permissions
- [ ] refactor permission service validation

### PR DESCRIPTION
**Pull request by clemens-msupply**
Closes #750 

- simplify program related permissions (main permission are just document permissions now but we could introduce an extra Program permission if needed)
- the graph_ql layer extracts a list of `allowed_docs` from the user permissions
- the service layer does the validation based on this list of `allowed_docs`
- update Document loaders, they now have to group requests by `allowed_docs`

Note: the `context_permissions` helper to get the `allowed_docs` list takes a permission parameter. However, so far we hid all the permissions in the DSL and having it not in the DSL is a bit against the point of having the DSL. (reason for the DSL was to have all permission requirements in one place...). I think I have a simple solution to solve this issue though, will create another PR with a minimal example...

### CODE REVIEW

**Comment by mark-prins** [comment]
If I remove `DOCUMENT_MUTATE` for the `Patient` context and try to create a patient, the f/e messaging  could be better:

Screenshot 2022-11-04 at 12 34 56 PM

The patient does belong to the store!

Also, after the error, I was taken to a patient record which had the first & last name which I'd entered. Clicking on the patient list view however.. my new patient is not shown 😕 This is all a little confusing..!

File: server/service/src/programs/encounter/insert.rs [hunk 1]
```diff
@@ -68,8 +73,11 @@ pub fn insert_encounter(
             // Updating the document will trigger an update in the patient (names) table
             let result = service_provider
                 .document_service
-                .update_document(ctx, doc)
+                .update_document(ctx, doc, &allowed_docs)
                 .map_err(|err| match err {
+                    DocumentInsertError::NotAllowedToMutDocument => {
```
Comment (mark-prins) [inline server/service/src/programs/encounter/insert.rs:78]: :nit I would prefer `NotAllowedToMutateDocument` as I don't think that `Mut` is a generally accepted abbreviation. Makes sense in a rust context.. but beyond that? 🤷

**Review by mark-prins — COMMENTED** [review]
😮  comprehensive! and much neater - changes look good!

Am having an issue testing - I had removed permission for patient document mutation, and received an error when saving. Now, I can create patients and save them fine?? Unable to see what's wrong. Will recreate the database again and retry..

Was wondering about f/e error messaging, and think that another issue to review that might be helpful.

**Comment by mark-prins** [comment]
Still unable to get an error when editing patients! 😢 

I do get some interesting things if I remove the `HIVCareEncounter` permission:

Screenshot 2022-11-04 at 1 17 24 PM

which then shows this if I dismiss:

Screenshot 2022-11-04 at 1 18 05 PM

File: server/graphql/programs/src/mutations/patient/insert.rs [hunk 2]
```diff
@@ -3,8 +3,9 @@ use graphql_core::{
     standard_graphql_error::{validate_auth, StandardGraphqlError},
     ContextExt,
 };
+use repository::Permission;
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/graphql/programs/src/mutations/patient/insert.rs:6]: +1 In the other PR I added the document permissions as a requirement for the patient endpoint in the DSL.

File: server/service/src/programs/encounter/insert.rs (hunk 1 above)
Comment (clemens-msupply) [reply server/service/src/programs/encounter/insert.rs]: thanks I changed it, Mut is fine for Rust code but the Mut was also used in graphql. I renamed all Mut to Mutate.

**Comment by roxy-dao** [comment]
Looks good (: and everything works~

But I can still see the patient list even when I don't have permission to view them and this error message pops up when I click on a patient to view their details/create a new patient. Should return the error message? 🤔 
Screen Shot 2022-11-07 at 11 45 25 AM

**Comment by clemens-msupply** [comment]
@mark-prins @roxy-dao I added another issue for the point you mentioned:
https://github.com/openmsupply/open-msupply/issues/766

**Pull request merged**